File: tests/cpp/unit/RuleRuntimeStateTests.cpp
#include "services/RuleRuntimeState.h"
#include "models/ProcessPriorityRule.h"

#include <gtest/gtest.h>

using namespace imeindicator;

namespace {

models::ProcessPriorityRule makeRule(int maxBackoff = 6)
{
    models::ProcessPriorityRule r;
    r.processName = L"chrome";
    r.targetPriority = models::PriorityLevel::BelowNormal;
    r.maxBackoffExponent = maxBackoff;
    r.isEnabled = true;
    r.useECoreOnly = false;
    return r;
}

} // namespace

TEST(RuleRuntimeStateTests, InitiallyDueAndExponentZero)
{
    auto rule = makeRule();
    services::RuleRuntimeState s(rule, 1);
    EXPECT_EQ(s.currentExponent(), 0);
    EXPECT_TRUE(s.isDue(services::RuleRuntimeState::clock::now()));
    EXPECT_FALSE(s.lastResult().has_value());
}

TEST(RuleRuntimeStateTests, BackoffIncreaseUpToMax)
{
    auto rule = makeRule(3);
    services::RuleRuntimeState s(rule, 1);
    auto now = services::RuleRuntimeState::clock::now();
    for (int i = 0; i < 10; ++i) s.backoffIncrease(now);
    EXPECT_EQ(s.currentExponent(), 3);
    EXPECT_EQ(s.lastResult().value(), services::MonitorResult::Skipped);
}

TEST(RuleRuntimeStateTests, ResetAfterChangeReturnsToZero)
{
    auto rule = makeRule(3);
    services::RuleRuntimeState s(rule, 1);
    auto now = services::RuleRuntimeState::clock::now();
    s.backoffIncrease(now);
    s.backoffIncrease(now);
    s.resetAfterChange(now);
    EXPECT_EQ(s.currentExponent(), 0);
    EXPECT_EQ(s.lastResult().value(), services::MonitorResult::Success);
}

TEST(RuleRuntimeStateTests, ResetAfterFailedKeepsResult)
{
    auto rule = makeRule(6);
    services::RuleRuntimeState s(rule, 1);
    auto now = services::RuleRuntimeState::clock::now();
    s.resetAfterSkipOrError(now, services::MonitorResult::Failed);
    EXPECT_EQ(s.currentExponent(), 0);
    EXPECT_EQ(s.lastResult().value(), services::MonitorResult::Failed);
}

TEST(RuleRuntimeStateTests, IntervalDoublesWithExponent)
{
    auto rule = makeRule(6);
    services::RuleRuntimeState s(rule, 1);
    auto now = services::RuleRuntimeState::clock::now();
    EXPECT_EQ(s.currentIntervalSeconds(), 1);
    s.backoffIncrease(now);
    EXPECT_EQ(s.currentIntervalSeconds(), 2);
    s.backoffIncrease(now);
    EXPECT_EQ(s.currentIntervalSeconds(), 4);
}

TEST(RuleRuntimeStateTests, UpdatePollingIntervalChangesEffectiveInterval)
{
    auto rule = makeRule(6);
    services::RuleRuntimeState s(rule, 1);
    EXPECT_EQ(s.currentIntervalSeconds(), 1);
    s.updatePollingInterval(5);
    EXPECT_EQ(s.currentIntervalSeconds(), 5);
}

File: tests/csharp/IMEIndicator.Tests/Services/RuleRuntimeStateTests.cs
// Copyright (C) 2026 IMEIndicator Project
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License v2 or later.
// See COPYING in the repository root for the full license text.

using IMEIndicator.Models;
using IMEIndicator.Services;
using Xunit;

namespace IMEIndicator.Tests.Services;

/// <summary>
/// <see cref="RuleRuntimeState"/> のテスト。現行 tests/cpp/unit/RuleRuntimeStateTests.cpp の移植。
/// </summary>
public sealed class RuleRuntimeStateTests
{
    private static ProcessPriorityRule MakeRule(int maxBackoff = 6) => new()
    {
        ProcessName = "chrome",
        TargetPriority = PriorityLevel.BelowNormal,
        MaxBackoffExponent = maxBackoff,
        IsEnabled = true,
        UseECoreOnly = false,
    };

    [Fact]
    public void InitiallyDueAndExponentZero()
    {
        RuleRuntimeState state = new(MakeRule(), 1);

        Assert.Equal(0, state.CurrentExponent);
        Assert.True(state.IsDue(RuleRuntimeState.Clock.Now));
        Assert.Null(state.LastResult);
    }

    [Fact]
    public void BackoffIncreaseUpToMax()
    {
        RuleRuntimeState state = new(MakeRule(3), 1);
        TimeSpan now = RuleRuntimeState.Clock.Now;

        for (int i = 0; i < 10; i++)
        {
            state.BackoffIncrease(now);
        }

        Assert.Equal(3, state.CurrentExponent);
        Assert.Equal(MonitorResult.Skipped, state.LastResult);
    }

    [Fact]
    public void ResetAfterChangeReturnsToZero()
    {
        RuleRuntimeState state = new(MakeRule(3), 1);
        TimeSpan now = RuleRuntimeState.Clock.Now;

        state.BackoffIncrease(now);
        state.BackoffIncrease(now);
        state.ResetAfterChange(now);

        Assert.Equal(0, state.CurrentExponent);
        Assert.Equal(MonitorResult.Success, state.LastResult);
    }

    [Fact]
    public void ResetAfterFailedKeepsResult()
    {
        RuleRuntimeState state = new(MakeRule(), 1);
        TimeSpan now = RuleRuntimeState.Clock.Now;

        state.ResetAfterSkipOrError(now, MonitorResult.Failed);

        Assert.Equal(0, state.CurrentExponent);
        Assert.Equal(MonitorResult.Failed, state.LastResult);
    }

    [Fact]
    public void IntervalDoublesWithExponent()
    {
        RuleRuntimeState state = new(MakeRule(), 1);
        TimeSpan now = RuleRuntimeState.Clock.Now;

        Assert.Equal(1, state.CurrentIntervalSeconds);
        state.BackoffIncrease(now);
        Assert.Equal(2, state.CurrentIntervalSeconds);
        state.BackoffIncrease(now);
        Assert.Equal(4, state.CurrentIntervalSeconds);
    }

    [Fact]
    public void UpdatePollingIntervalChangesEffectiveInterval()
    {
        RuleRuntimeState state = new(MakeRule(), 1);

        Assert.Equal(1, state.CurrentIntervalSeconds);
        state.UpdatePollingInterval(5);
        Assert.Equal(5, state.CurrentIntervalSeconds);
    }
}
